File: expectanim/src/main/java/com/github/florent37/expectanim/core/position/PositionAnimExpectationCenterBetweenViewAndParent.java
package com.github.florent37.expectanim.core.position;

import android.view.View;
import android.view.ViewParent;

/**
 * Created by florentchampigny on 17/02/2017.
 */

public class PositionAnimExpectationCenterBetweenViewAndParent extends PositionAnimationViewDependant {

    private boolean horizontal;
    private boolean vertical;
    private boolean toBeOnRight;
    private boolean toBeOnBottom;

    public PositionAnimExpectationCenterBetweenViewAndParent(View otherView, boolean horizontal, boolean vertical, boolean toBeOnRight, boolean toBeOnBottom) {
        super(otherView);
        this.horizontal = horizontal;
        this.vertical = vertical;
        this.toBeOnRight = toBeOnRight;
        this.toBeOnBottom = toBeOnBottom;

        setForPositionY(true);
        setForPositionX(true);
    }

    @Override
    public Float getCalculatedValueX(View viewToMove) {
        if(horizontal){
            final ViewParent viewParent = otherView.getParent();
            if((viewParent instanceof View) && horizontal){
                final View parentView = (View) viewParent;
                final float centerOfOtherView = viewCalculator.finalCenterXOfView(otherView);
                if(toBeOnRight) {
                    final float parentWidth = parentView.getWidth();
                    return (parentWidth + centerOfOtherView) / 2f - viewToMove.getWidth() / 2f;
                } else {
                    return centerOfOtherView / 2f - viewToMove.getWidth() / 2f;
                }
            }
        }
        return null;
    }

    @Override
    public Float getCalculatedValueY(View viewToMove) {
        if(vertical){
            final ViewParent viewParent = viewToMove.getParent();
            if((viewParent instanceof View) && vertical){
                final View parentView = (View)viewParent;
                final float centerOfOtherView = viewCalculator.finalCenterYOfView(otherView);
                if(toBeOnBottom){
                    final float parentHeight = parentView.getHeight();
                    return parentHeight + centerOfOtherView / 2f - viewToMove.getHeight() / 2f;
                } else {
                    return centerOfOtherView / 2f - viewToMove.getHeight() / 2f;
                }

            }
        }
        return null;
    }
}

File: kotlinpleaseanimate/src/main/java/com/github/florent37/kotlin/pleaseanimate/core/position/PositionAnimExpectationCenterBetweenViewAndParent.kt
package com.github.florent37.kotlin.pleaseanimate.core.position

import android.view.View

class PositionAnimExpectationCenterBetweenViewAndParent(otherView: View, private val horizontal: Boolean, private val vertical: Boolean, private val toBeOnRight: Boolean, private val toBeOnBottom: Boolean) : PositionAnimationViewDependant(otherView) {

    init {
        isForPositionY = true
        isForPositionX = true
    }

    override fun getCalculatedValueX(viewToMove: View): Float? {
        if (horizontal) {
            val viewParent = otherView.parent
            if (viewParent is View && horizontal) {
                val parentView = viewParent as View
                val centerOfOtherView = viewCalculator!!.finalCenterXOfView(otherView)
                if (toBeOnRight) {
                    val parentWidth = parentView.width.toFloat()
                    return (parentWidth + centerOfOtherView) / 2f - viewToMove.width / 2f
                } else {
                    return centerOfOtherView / 2f - viewToMove.width / 2f
                }
            }
        }
        return null
    }

    override fun getCalculatedValueY(viewToMove: View): Float? {
        if (vertical) {
            val viewParent = viewToMove.parent
            if (viewParent is View && vertical) {
                val parentView = viewParent as View
                val centerOfOtherView = viewCalculator!!.finalCenterYOfView(otherView)
                if (toBeOnBottom) {
                    val parentHeight = parentView.height.toFloat()
                    return parentHeight + centerOfOtherView / 2f - viewToMove.height / 2f
                } else {
                    return centerOfOtherView / 2f - viewToMove.height / 2f
                }
            }
        }
        return null
    }
}
